tests/mbt/driver/mbt_test.go

```diff
@@ -0,0 +1,577 @@
+				driver.delegate(int64(valIndex), changeAmount)
+			} else {
+				// undelegate from the validator
+				driver.undelegate(int64(valIndex), -changeAmount)
+			}
+
+			stats.numTxs++
+		case "EndAndBeginBlockForProvider":
+			timeAdvancement := lastAction["timeAdvancement"].Value.(int64)
+			consumersToStart := lastAction["consumersToStart"].Value.(itf.ListExprType)
+			consumersToStop := lastAction["consumersToStop"].Value.(itf.ListExprType)
+			t.Log("EndAndBeginBlockForProvider", timeAdvancement, consumersToStart, consumersToStop)
+
+			stats.numStartedChains += len(consumersToStart)
+			stats.numStops += len(consumersToStop)
+
+			// we need 2 blocks, because for a packet sent at height H, the receiving chain
+			// needs a header of height H+1 to accept the packet
+			// so we do one time advancement with a very small increment,
+			// and then increment the rest of the time]
+			runningConsumersBefore := driver.runningConsumers()
+			driver.endAndBeginBlock("provider", 1*time.Nanosecond)
+			driver.endAndBeginBlock("provider", time.Duration(timeAdvancement)*time.Second-1*time.Nanosecond)
+			runningConsumersAfter := driver.runningConsumers()
+
+			// the consumers that were running before but not after must have timed out
+			for _, consumer := range runningConsumersBefore {
+				found := false
+				for _, consumerAfter := range runningConsumersAfter {
+					if consumerAfter.ChainId == consumer.ChainId {
+						found = true
+						break
+					}
+				}
+				if !found {
+					stats.numTimeouts++
+				}
+			}
+
+			// save the last timestamp for runningConsumers,
+			// because setting up chains will modify timestamps
+			// when the coordinator is starting chains
+			lastTimestamps := make(map[ChainId]time.Time, len(consumers))
+			for _, consumer := range driver.runningConsumers() {
+				lastTimestamps[ChainId(consumer.ChainId)] = driver.runningTime(ChainId(consumer.ChainId))
+			}
+
+			driver.coordinator.CurrentTime = driver.runningTime("provider")
+			// start consumers
+			for _, consumer := range consumersToStart {
+				driver.setupConsumer(
+					consumer.Value.(string),
+					modelParams,
+					driver.providerChain().Vals,
+					signers,
+					nodes,
+					valNames,
+					driver.providerChain(),
+				)
+			}
+
+			// stop consumers
+			for _, consumer := range consumersToStop {
+				driver.stopConsumer(ChainId(consumer.Value.(string)))
+			}
+
+			// reset the times for the consumers that were not stopped or started just now
+			// their times were messed up by the coordinator
+			for consumer, timestamp := range lastTimestamps {
+				driver.setTime(consumer, timestamp)
+			}
+
+			// for all connected consumers, update the clients...
+			// unless it was the last consumer to be started, in which case it already has the header
+			for _, consumer := range driver.runningConsumers() {
+				if len(consumersToStart) > 0 && consumer.ChainId == consumersToStart[len(consumersToStart)-1].Value.(string) {
+					continue
```

Comment:
It's not clear to me why the last consumer to be started already has the provider's header?